src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1514,37 +1536,53 @@ class Simplex_tree {
   }
 
  public:
-  /** \brief This function ensures that each simplex has a higher filtration value than its faces by increasing the
-   * filtration values.
-   * @return True if any filtration value was modified, false if the filtration was already non-decreasing.
-   * 
-   * If a simplex has a `NaN` filtration value, it is considered lower than any other defined filtration value.
+  /** Calls a function on each simplex. The order ensures that faces are visited before cofaces.
+   * While it is fine to modify the data of a simplex (filtration, key) in the function, modifying
+   * the structure itself (insertion, removal) is not supported.
+   *
+   * @param[in] fun Function that takes as argument a Simplex_handle and an int (representing the dimension of this
+   * simplex). It may return void or bool, and in the second case returning true means that the iteration will skip
+   * the children of this simplex (a subset of the cofaces).
    */
-  bool make_filtration_non_decreasing() {
-    bool modified = false;
-    // Loop must be from the end to the beginning, as higher dimension simplex are always on the left part of the tree
-    for (auto& simplex : boost::adaptors::reverse(root_.members())) {
-      if (has_children(&simplex)) {
-        modified |= rec_make_filtration_non_decreasing(simplex.second.children());
+  template<class Fun>
+  void for_each_simplex(Fun&& fun) {
```

Comment:
I have seen your comment about the method name, and I would prefer `for_each_simplex_with_dim` or `for_each_simplex_by_dim`. But it is only a matter of taste, it is ok to me if you let this name.